README.rst
rabbit_force
============

.. image:: https://readthedocs.org/projects/rabbit-force/badge/?version=latest
    :target: https://rabbit-force.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status

.. image:: https://travis-ci.org/robertmrk/rabbit_force.svg?branch=develop
    :target: https://travis-ci.org/robertmrk/rabbit_force
    :alt: Build status

.. image:: https://coveralls.io/repos/github/robertmrk/rabbit_force/badge.svg
    :target: https://coveralls.io/github/robertmrk/rabbit_force
    :alt: Coverage

.. image:: https://img.shields.io/badge/License-MIT-yellow.svg
    :target: https://opensource.org/licenses/MIT
    :alt: MIT license

rabbit_force is a `Salesforce Streaming API <api_>`_ to `RabbitMQ <rabbitmq_>`_
adapter service. It listens for event messages from
`Salesforce's Streaming API <api_>`_ and forwards them to a
`RabbitMQ broker <rabbitmq_>`_ for you, so you don't have to.

`Streaming API <api_>`_ is useful when you want notifications to be pushed from
the server to the client based on criteria that you define with
`PushTopics <PushTopic_>`_ or to receive
`generic streaming <GenericStreaming_>`_ messages. While
`RabbitMQ <rabbitmq_>`_ is one of the most popular options for implementing
inter-service communication in a `microservice architecture <microservice_>`_.

While there are lots of great client implementations for RabbitMQ/AMQP for
various languages, there are a lot less Streaming API clients, and many of them
are badly maintained. Furthermore RabbitMQ offers much more flexible message
consumption techniques.

rabbit_force aims to fix these problems, by providing and adapter between the
Streaming API and RabbitMQ, so inter-connected services can consume Streaming
API event messages with RabbitMQ. It even supports connection with multiple
Salesforce orgs, multiple RabbitMQ brokers, and routing messages between them.

Features
--------

- Forward `Streaming API <api_>`_ messages from one or more Salesforce orgs to
  one or more `RabbitMQ brokers <rabbitmq_>`_
- Route incoming messages to a specific broker and exchange with the
  specified routing key and properties, with the help of routing rules defined
  as `JSONPath <jsonpath_>`_ expressions
- Support for Salesforce's replay extension for `message reliability and
  durability <replay_>`_ by storing replay markers in a `Redis <redis_>`_
  database
- Configurable error handling behavior, either fail instantly or try to recover
  from network and service outages
- Message sources, sinks and routing configurable with JSON or YAML
  configuration files
- Implemented using `python asyncio <asyncio_>`_ for efficient handling of
  IO intensive operations

Usage
-----

A relatively simple use case is illustrated on the image bellow. rabbit_force
is connected to a single message source and message sink, or in other words to
a single Salesforce org named as ``my_org`` and a single RabbitMQ broker named
``my_broker``. It listens for messages from two PushTopics (``lead_changes``
and ``contact_changes``) and a StreamingChannel (``my_channel``), and forwards
messages into the exchange ``my_exchange`` with different routing keys. A redis
database is used to store replay markers sent by Salesforce to take advantage
of `message durability <replay_>`_.

.. image:: docs/source/_static/usage.svg

The configuration file bellow sets up rabbit_force to forward messages from
``lead_changes``, ``contact_changes`` and ``my_channel`` with the routing keys
of ``lead_change_message``, ``contact_change_message`` and
``my_channel_message`` respectively.

config.yaml:

.. code-block:: yaml

    # message source definition
    source:
      # mapping of Salesforce orgs to use as message sources
      orgs:
        # a Salesforce org named as "my_org"
        my_org:
          # authentication credentials
          consumer_key: "<consumer_key>"
          consumer_secret: "<consumer_secret>"
          username: "<username>"
          password: "<password>"
          # list of resources that the service can listen to for messages
          # if they doesn't exist, they'll be created on application startup
          resources:
            # a PushTopic resource
            - type: PushTopic
              # the definition of the PushTopic
              spec:
                Name: lead_changes
                ApiVersion: 42.0
                NotifyForFields: Referenced
                NotifyForOperationCreate: true
                NotifyForOperationUpdate: true
                NotifyForOperationDelete: true
                NotifyForOperationUndelete: true
                Query: SELECT Id, Email, Name, Phone, MobilePhone, Status, LeadSource FROM Lead
              # optional durable flag, if false then the resource will be removed on application shutdown
              durable: false

            # a PushTopic resource
            - type: PushTopic
              # the definition of the PushTopic
              spec:
                Name: contact_changes
                ApiVersion: 42.0
                NotifyForFields: Referenced
                NotifyForOperationCreate: true
                NotifyForOperationUpdate: true
                NotifyForOperationDelete: true
                NotifyForOperationUndelete: true
                Query: SELECT Id, Email, Name, Phone, MobilePhone FROM Contact
              # optional durable flag, if false then the resource will be removed on application shutdown
              durable: false

            # a StreamingChannel resource
            - type: StreamingChannel
              # the definition of the StreamingChannel
              spec:
                Name: /u/my_channel
                Description: Streaming channel for notifications
      # optional replay storage definition. if defined it'll be used to store replay
      # markers sent by Salesforce in order to support message durability
      replay:
        # redis server address
        address: "redis://localhost:6389"
        # key prefix
        key_prefix: replay

    # message sink definition
    sink:
      # mapping of RabbitMQ brokers to use as message sinks
      brokers:
        # a RabbitMQ broker named as "by_broker"
        my_broker:
          # host name of the broker
          host: localhost
          # definition of the exchange where the messages should be forwarded
          exchanges:
            - exchange_name: my_exchange
              type_name: topic
              durable: true

    # message router definition
    router:
      # optional default route to use if no routing rule matches a given message
      default_route:
        broker_name: my_broker
        exchange_name: my_exchange
        routing_key: my_channel_message

      # list of routing rules
      rules:
        # JSONPath filter expression as the condition
        - condition: "$[?(@.message.channel ~ '.*/lead_changes')]"
          # the route to use if the condition produces a non-empty match
          route:
            broker_name: my_broker
            exchange_name: my_exchange
            routing_key: lead_change_message
        # JSONPath filter expression as the condition
        - condition: "$[?(@.message.channel ~ '.*/contact_changes')]"
          # the route to use if the condition produces a non-empty match
          route:
            broker_name: my_broker
            exchange_name: my_exchange
            routing_key: contact_change_message


A sample run of rabbit_force with the above configuration file.

.. code-block:: bash

    $ python -m rabbit_force config.yaml
    2018-06-19 16:23:07,909:INFO: Starting up ...
    2018-06-19 16:23:07,996:INFO: Configuration loaded from 'config.yaml'
    2018-06-19 16:23:07,999:INFO: Configuring application ...
    2018-06-19 16:23:10,619:INFO: Using message broker AmqpBroker(host='localhost', port=None, login='guest', password='guest', virtualhost='/', ssl=False, login_method='AMQPLAIN', insist=False, verify_ssl=True)
    2018-06-19 16:23:12,128:INFO: Listening for messages from Salesforce org 'my_org':
        * from PushTopic 'lead_changes' on channel '/topic/lead_changes'
        * from PushTopic 'contact_changes' on channel '/topic/contact_changes'
        * from StreamingChannel '/u/my_channel' on channel '/u/my_channel'
    With replay storage RedisReplayStorage(address='redis://localhost:6389', key_prefix='replay:my_org', additional_params={}, ignore_network_errors=False).
    2018-06-19 16:23:48,119:INFO: Forwarded message 1 on channel '/topic/lead_changes' from 'my_org' to Route(broker_name='my_broker', exchange_name='my_exchange', routing_key='lead_change_message', properties=None).
    2018-06-19 16:24:03,039:INFO: Forwarded message 1 on channel '/topic/contact_changes' from 'my_org' to Route(broker_name='my_broker', exchange_name='my_exchange', routing_key='contact_change_message', properties=None).
    2018-06-19 16:24:20,180:INFO: Forwarded message 1 on channel '/u/my_channel' from 'my_org' to Route(broker_name='my_broker', exchange_name='my_exchange', routing_key='my_channel_message', properties=None).
    2018-06-19 16:24:27,097:INFO: Shutting down ...

More advanced examples can be found in the `examples section of the
documentation <example_docs_>`_.

Documentation
-------------

`https://rabbit-force.readthedocs.io <https://rabbit-force.readthedocs.io>`_

Requirements
------------

Python 3.7

Docker
------

The `public docker image <https://hub.docker.com/r/robertmrk/rabbit_force>`_ can
be found on docker hub.

.. code-block:: bash

    $ docker pull robertmrk/rabbit_force

License
-------

rabbit_force is offered under the `MIT license <LICENSE.txt>`_.

.. _asyncio: https://docs.python.org/3/library/asyncio.html
.. _api: https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/intro_stream.htm
.. _PushTopic: https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/working_with_pushtopics.htm
.. _GenericStreaming: https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/generic_streaming_intro.htm#generic_streaming_intro
.. _replay: https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/using_streaming_api_durability.htm
.. _rabbitmq: http://www.rabbitmq.com/
.. _microservice: http://microservices.io/patterns/communication-style/messaging.html
.. _jsonpath: http://goessner.net/articles/JsonPath/
.. _redis: https://redis.io/
.. _example_docs: https://rabbit-force.readthedocs.io/en/latest/examples.html


CHANGELOG.rst
Changelog
=========

0.1.1 (2019-05-15)
------------------

- Switch to python 3.7
- Update outdated dependencies
- Add docker deployment option


0.1.0 (2018-07-05)
------------------

- Forward `Streaming API <api_>`_ messages from one or more Salesforce orgs to
  one or more `RabbitMQ brokers <rabbitmq_>`_
- Route incoming messages to a specific broker and exchange with the
  specified routing key and properties, with the help of routing rules defined
  as `JSONPath <jsonpath_>`_ expressions
- Support for Salesforce's replay extension for `message reliability and
  durability <replay_>`_ by storing replay markers in a `Redis <redis_>`_
  database
- Configurable error handling behavior, either fail instantly or try to recover
  from network and service outages
- Message sources, sinks and routing configurable with JSON or YAML
  configuration files
- Implemented using `python asyncio <asyncio_>`_ for efficient handling of
  IO intensive operations

.. _aiohttp: https://github.com/aio-libs/aiohttp/
.. _aiocometd: https://github.com/robertmrk/aiocometd/
.. _asyncio: https://docs.python.org/3/library/asyncio.html
.. _api: https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/intro_stream.htm
.. _PushTopic: https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/working_with_pushtopics.htm
.. _GenericStreaming: https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/generic_streaming_intro.htm#generic_streaming_intro
.. _replay: https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/api_streaming/using_streaming_api_durability.htm
.. _CometD: https://cometd.org/
.. _Comet: https://en.wikipedia.org/wiki/Comet_(programming)
.. _Bayeux: https://docs.cometd.org/current/reference/#_bayeux
.. _ext: https://docs.cometd.org/current/reference/#_bayeux_ext
.. _password_auth: https://help.salesforce.com/articleView?id=remoteaccess_oauth_username_password_flow.htm&type=5
.. _refresh_auth: https://help.salesforce.com/articleView?id=remoteaccess_oauth_refresh_token_flow.htm&type=5
.. _connected_app: https://help.salesforce.com/articleView?id=connected_app_overview.htm&type=5
.. _sf_auth: https://help.salesforce.com/articleView?id=remoteaccess_authenticate_overview.htm
.. _web_server_auth: https://help.salesforce.com/articleView?id=remoteaccess_oauth_web_server_flow.htm&type=5
.. _rabbitmq: http://www.rabbitmq.com/
.. _microservice: http://microservices.io/patterns/communication-style/messaging.html
.. _jsonpath: http://goessner.net/articles/JsonPath/
.. _redis: https://redis.io/


rabbit_force/config.py
"""Marshmallow schemas for configuration validation and functions for
loading configurations from files"""
from pathlib import Path
import json
import logging

from marshmallow import Schema, fields, validates_schema, ValidationError, \
    post_load, RAISE
from marshmallow.validate import Length, Range, OneOf
import yaml

from .salesforce import StreamingResourceType
from .exceptions import ConfigurationError


LOGGER = logging.getLogger(__name__)


class StrictSchema(Schema):
    """Common schema base class which rejects unknown fields"""
    def __init__(self, *args, **kwargs):
        kwargs["unknown"] = RAISE
        super().__init__(*args, **kwargs)


class PushTopicSchema(StrictSchema):
    """Configuration schema for PushTopic resources"""

    # PushTopic fields are validated according to
    # https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/\
    # api_streaming/pushtopic.htm
    Id = fields.String(validate=Length(min=1))
    Name = fields.String(validate=Length(min=1, max=25))
    ApiVersion = fields.Float(validate=Range(min=20.0, max=42.0))
    IsActive = fields.Boolean(default=True)
    NotifyForFields = fields.String(validate=OneOf(["All",
                                                    "Referenced",
                                                    "Select",
                                                    "Where"]))
    Description = fields.String(default=None, validate=Length(max=400))
    NotifyForOperationCreate = fields.Boolean(default=True)
    NotifyForOperationUpdate = fields.Boolean(default=True)
    NotifyForOperationDelete = fields.Boolean(default=True)
    NotifyForOperationUndelete = fields.Boolean(default=True)
    NotifyForOperations = fields.String(validate=OneOf(["All",
                                                        "Create",
                                                        "Extended",
                                                        "Update"]))
    Query = fields.String(validate=Length(min=1, max=1300))

    @validates_schema
    def check_required_fileds(self, data):  # pylint: disable=no-self-use
        """Check for required fields

        :raise marshmallow.exceptions.ValidationError: If no fields are \
        specified or if only a single non identifier field is specified or \
        multiple fields are specified but they're not enough for a resource \
        definition
        """
        if len(data) == 1:
            unique_id_fields = {"Id", "Name"}
            if not data.keys() & unique_id_fields:
                raise ValidationError("If only a single field is specified "
                                      "it should be a unique identifier like "
                                      "'Id' or 'Name'.")
        elif len(data) > 1:
            required_fields = {"Name", "ApiVersion", "Query"}
            if (data.keys() & required_fields) != required_fields:
                raise ValidationError("If multiple fields are specified it "
                                      "it should be a full resource "
                                      "definition where at least 'Name', "
                                      "'ApiVersion' and 'Query' are required.")
        else:
            raise ValidationError("Either a single fields should be specified "
                                  "which uniquely identifies the resource or "
                                  "multiple fields which can be used to "
                                  "construct the resource.")

    @validates_schema
    def check_api_version(self, data):  # pylint: disable=no-self-use
        """Check for invalid fields for the specified API version

        :raise marshmallow.exceptions.ValidationError: If any invalid fields \
        found for the specified API version
        """
        # skip validation if the ApiVersion field is not present, which might
        # happen even when it's specified but it's value fails on validation
        if "ApiVersion" not in data:
            return

        # check for the presence of old fields for a newer API version
        if (data["ApiVersion"] >= 29.0 and
                "NotifyForOperations" in data):
            raise ValidationError("'NotifyForOperations' can only be specified"
                                  " for API version 28.0 and earlier.")

        # check for the presence of new fields for an older API version
        if (data["ApiVersion"] <= 28.0 and
                ("NotifyForOperationCreate" in data or
                 "NotifyForOperationDelete" in data or
                 "NotifyForOperationUndelete" in data or
                 "NotifyForOperationUpdate" in data)):
            raise ValidationError("'NotifyForOperationCreate', "
                                  "'NotifyForOperationDelete', "
                                  "'NotifyForOperationUndelete' and "
                                  "'NotifyForOperationUpdate' can only be "
                                  "specified for API version 29.0 and later.")


class StreamingChannelSchema(StrictSchema):
    """Configuration schema for StreamingChannel resources"""

    # StreamingChannel fields are validated according to
    # https://developer.salesforce.com/docs/atlas.en-us.api_streaming.meta/\
    # api_streaming/streamingChannel.htm
    Id = fields.String(validate=Length(min=1))
    Name = fields.String(validate=Length(min=1, max=80))
    Description = fields.String(default=None, validate=Length(max=255))

    @validates_schema
    def check_required_fileds(self, data):  # pylint: disable=no-self-use
        """Check for required fields

        :raise marshmallow.exceptions.ValidationError: If no fields are \
        specified or if only a single non identifier field is specified or \
        multiple fields are specified but they're not enough for a resource \
        definition
        """
        if len(data) == 1:
            unique_id_fields = {"Id", "Name"}
            if not data.keys() & unique_id_fields:
                raise ValidationError("If only a single field is specified "
                                      "it should be a unique identifier like "
                                      "'Id' or 'Name'.")
        elif not data:
            raise ValidationError("Either a single fields should be specified "
                                  "which uniquely identifies the resource or "
                                  "multiple fields which can be used to "
                                  "construct the resource.")


class StreamingResourceSchema(StrictSchema):
    """Configuration schema for streaming resources"""

    type = fields.String(required=True, attribute="resource_type",
                         validate=OneOf([_.value for _ in
                                         StreamingResourceType]))
    spec = fields.Dict(required=True, attribute="resource_spec")
    durable = fields.Boolean()

    @post_load
    def load_spec(self, data):
        """Load the spec field with the appropriate schema based on the
        type field"""
        # resource type scpecific schema classes
        schema_map = {
            StreamingResourceType.PUSH_TOPIC: PushTopicSchema,
            StreamingResourceType.STREAMING_CHANNEL: StreamingChannelSchema
        }

        # get the resource type value
        type_name = data[self.fields["type"].attribute]

        # get the schema class for the resource type
        schema_cls = schema_map[type_name]

        # load and update the value of spec field
        spec = data[self.fields["spec"].attribute]
        data[self.fields["spec"].attribute] = schema_cls().load(spec)

        # return the updated data
        return data


class SalesforceOrgSchema(StrictSchema):
    """Configuration schema for a Salesforce organization"""
    consumer_key = fields.String(required=True)
    consumer_secret = fields.String(required=True)
    username = fields.String(required=True)
    password = fields.String(required=True)
    resources = fields.List(fields.Nested(StreamingResourceSchema()),
                            required=True,
                            validate=Length(min=1),
                            attribute="streaming_resource_specs")


class ReplaySchema(StrictSchema):
    """Configuration schema for a Redis replay marker storage"""
    address = fields.Url(schemes=("redis",), required=True)
    key_prefix = fields.String()


class MessageSourceSchema(StrictSchema):
    """Configuration schema for a message source"""
    orgs = fields.Dict(keys=fields.String(),
                       values=fields.Nested(SalesforceOrgSchema()),
                       required=True,
                       validate=Length(min=1),
                       attribute="org_specs")
    replay = fields.Nested(ReplaySchema(), attribute="replay_spec")


class AmqpExchangeSchema(StrictSchema):
    """Configuration schema for declaring AMQP exchanges"""
    exchange_name = fields.String(required=True, validate=Length(min=1))
    type_name = fields.String(required=True, validate=OneOf(["fanout",
                                                             "direct",
                                                             "topic",
                                                             "headers"]))
    passive = fields.Boolean(default=False)
    durable = fields.Boolean(default=False)
    auto_delete = fields.Boolean(default=False)
    no_wait = fields.Boolean(default=False)
    arguments = fields.Dict(allow_none=True)


class AmqpBrokerSchema(StrictSchema):
    """Configuration schema for AMQP connection parameters"""
    host = fields.String(required=True, validate=Length(min=1))
    port = fields.Int(default=None, allow_none=True,
                      validate=Range(min=1, max=(2 ** 16) - 1))
    login = fields.String(default="guest")
    password = fields.String(default="guest")
    virtualhost = fields.String(default="/")
    ssl = fields.Boolean(default=False)
    verify_ssl = fields.Boolean(default=True)
    login_method = fields.String(default="AMQPLAIN")
    insist = fields.Boolean(default=False)
    exchanges = fields.List(fields.Nested(AmqpExchangeSchema()),
                            required=True,
                            validate=Length(min=1),
                            attribute="exchange_specs")


class MessageSinkSchema(StrictSchema):
    """Configuration schema for a message sink"""
    brokers = fields.Dict(keys=fields.String(),
                          values=fields.Nested(AmqpBrokerSchema()),
                          required=True,
                          validate=Length(min=1),
                          attribute="broker_specs")


class RouteSchema(StrictSchema):
    """Configuration schema for route parameters"""
    broker_name = fields.String(required=True, validate=Length(min=1))
    exchange_name = fields.String(required=True)
    routing_key = fields.String(required=True, validate=Length(min=1))
    properties = fields.Dict(
        keys=fields.String(
            validate=OneOf(("content_type",
                            "content_encoding",
                            "headers",
                            "delivery_mode",
                            "priority",
                            "correlation_id",
                            "reply_to",
                            "expiration",
                            "message_id",
                            "timestamp",
                            "type",
                            "user_id",
                            "app_id",
                            "cluster_id"))
        ),
        allow_none=True
    )


class RoutingRuleSchema(StrictSchema):
    """Configuration schema for routing rule"""
    condition = fields.String(required=True, validate=Length(min=1),
                              attribute="condition_spec")
    route = fields.Nested(RouteSchema(), required=True, attribute="route_spec")


class MessageRouterSchema(StrictSchema):
    """Configuration schema for message router"""
    default_route = fields.Nested(RouteSchema(),
                                  required=True,
                                  allow_none=True,
                                  attribute="default_route_spec")
    rules = fields.List(fields.Nested(RoutingRuleSchema()),
                        missing=list,
                        attribute="rule_specs")


class ApplicationConfigSchema(StrictSchema):
    """Congiguration schema for setting up the complete rabbit_force
    application"""
    source = fields.Nested(MessageSourceSchema(), required=True)
    sink = fields.Nested(MessageSinkSchema(), required=True)
    router = fields.Nested(MessageRouterSchema(), required=True)


def get_config_loader(file_path):
    """Find the appropriate config loader for *file_path*

    :param str file_path: An absolute or relative file path
    :return: A callable capable of loading the config file
    :rtype: :func:`callable` or None
    """
    # get the file's suffix without the period
    path = Path(file_path)
    suffix = path.suffix.lower()
    if suffix:
        suffix = suffix[1:]

    # return the json or the yaml load function
    if suffix == "json":
        LOGGER.debug("Using JSON config loader for %r", file_path)
        return json.load
    elif suffix in ("yml", "yaml"):
        LOGGER.debug("Using YAML config loader for %r", file_path)
        return yaml.safe_load

    # if the suffix of the file is not recognized return None
    LOGGER.debug("No config loader found for %r", file_path)
    return None


def load_config(file_path):
    """Load and validate the application's configuration from *file_path*

    :param str file_path: An absolute or relative file path
    :return: The loaded configuration
    :rtype: dict
    :raise ConfigurationError: If the file_path's suffix is not recognized, \
    if the file can't be loaded or if validation of the file's contents fail.
    """
    # get the appropriate loader for the file_path
    loader = get_config_loader(file_path)
    if not loader:
        raise ConfigurationError(f"Unrecognized configuration file "
                                 f"format for {file_path!r}.")

    # load the config file's contents
    LOGGER.debug("Loading configuration from %r", file_path)
    try:
        with open(file_path, "rt") as file:
            unvalidated_config = loader(file)
    except Exception as error:
        raise ConfigurationError(f"Failed to load configuration "
                                 f"file {file_path!r}. {error!s}") from error

    # validate the contents of the file
    LOGGER.debug("Validating configuration")
    try:
        config = ApplicationConfigSchema().load(unvalidated_config)
    except ValidationError as error:
        raise ConfigurationError(f"Failed to validate configuration "
                                 f"file {file_path!r}. {error!s}") from error

    return config
